### PR TITLE
Add base functionality of owned entities

### PR DESCRIPTION
**Pull request by ILoveMyTortoise**

### CODE REVIEW

File: src/Nox.Types.EntityFramework/EntityBuilderAdapter/IEntityBuilderAdapter.cs
Comment (rochar) [inline src/Nox.Types.EntityFramework/EntityBuilderAdapter/IEntityBuilderAdapter.cs:1]: rename to IEntityBuilder
The implementations will be adapter

File: src/Nox.Types.EntityFramework/EntityBuilderAdapter/EntityBuilderAdapter.cs
Comment (rochar) [inline src/Nox.Types.EntityFramework/EntityBuilderAdapter/EntityBuilderAdapter.cs:1]: we need to split this in EntityTypeBuilderAdapter and OwnedNavigationBuilderAdapter
each is a different adapter and also decrease cyclomatic complexity

File: src/Nox.Generator/Application/Dto/EntityDto.template.cs [hunk 1]
```diff
@@ -11,6 +11,7 @@
 using Nox.Domain;
 using {{codeGeneratorState.DataTransferObjectsNameSpace}};
 using {{codeGeneratorState.DomainNameSpace}};
+using Microsoft.OData.ModelBuilder;
```
Comment (rochar) [inline src/Nox.Generator/Application/Dto/EntityDto.template.cs:14]: this using is strange? EntityDto should not dependn on Odata?

File: src/Nox.Solution/NoxSolution.cs [hunk 2]
```diff
@@ -39,6 +39,53 @@ internal void Validate()
...
+                    Relationship = relationship,
+                    RelationshipEntityType = getTypeByNameFunc(relationship.Entity)!
+                });
+            }
+        }
+
+        return fullRelationshipModels;
+    }
+
+    public bool IsOwnedEntity(Entity entity)
```
Comment (rochar) [inline src/Nox.Solution/NoxSolution.cs:63]: this method should be internal, the entity has a property that defines if is owned or not

everyone should use that

File: src/Nox.Solution/NoxSolution.cs [hunk 3]
```diff
@@ -39,6 +39,53 @@ internal void Validate()
...
+    {
+        if (Domain == null)
+        {
+            return false;
+        }
+
+        foreach(var domainEntity in Domain.Entities)
```
Comment (rochar) [inline src/Nox.Solution/NoxSolution.cs:70]: improve performance by  having a hashset of OwnedRelationships Entity name, and the use it for search

the hashset can be a input of the method calculate in EvaluateCalculatedProperties a single time (no local state is needed)

File: src/Nox.Solution/NoxSolutionBuilder.cs [hunk 4]
```diff
@@ -189,11 +189,24 @@ private NoxSolution ResolveAndLoadConfiguration()
...
             var config = NoxSchemaValidator.Deserialize<NoxSolution>(yaml);
 
+            EvaluateCalculatedProperties(config);
+
             config.RootYamlFile = _yamlFilePath;
 
             return config;
         }
 
+        private void EvaluateCalculatedProperties(NoxSolution config)
```
Comment (rochar) [inline src/Nox.Solution/NoxSolutionBuilder.cs:199]: I would prefer to be explicit ComputeIsOwnedEntityForAll()

File: src/Nox.Solution/NoxSolution.cs [hunk 5]
```diff
@@ -39,6 +39,53 @@ internal void Validate()
         return fullRelationshipModels;
     }
 
+    public List<EntityRelationshipWithType> GetOwnedRelationshipsToCreate(
+        Func<string, Type?> getTypeByNameFunc,
```
Comment (rochar) [inline src/Nox.Solution/NoxSolution.cs:43]: now looking at this, it seems to be an obscured thin

lets just rename to getEntityTypeByName we need to think how to improve this

File: src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs [hunk 6]
```diff
@@ -1,9 +1,14 @@
...
 
 public interface INoxDatabaseConfigurator
 {
-    void ConfigureEntity(NoxSolutionCodeGeneratorState codeGeneratorState, EntityTypeBuilder builder, Entity entity, IReadOnlyList<EntityRelationshipWithType> relationshipsToCreate);
+    void ConfigureEntity(
+        NoxSolutionCodeGeneratorState codeGeneratorState,
+        IEntityBuilderAdapter builder,
+        Entity entity,
+        NoxSolution noxSolution,
+        Func<string, Type?> getTypeByNameFunc);
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs:13]: lets discuss this Func<string, Type?> getTypeByNameFunc
is kind of obscure, we need to think about this

File: src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs [hunk 7]
```diff
@@ -107,8 +113,43 @@ public abstract class NoxDatabaseConfigurator : INoxDatabaseConfigurator
...
         }
 
-        private void ConfigureRelationForeignKeyProperty(NoxSolutionCodeGeneratorState codeGeneratorState,
-            EntityTypeBuilder builder,
+        public virtual void ConfigureOwnedRelationships(
+            NoxSolutionCodeGeneratorState codeGeneratorState,
+            IEntityBuilderAdapter builder,
+            Entity entity,
+            IReadOnlyList<EntityRelationshipWithType> ownedRelationshipsToCreate,
+            NoxSolution noxSolution,
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs:121]: the codeGeneratorState already contains a Solution no need to input it again

File: src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs [hunk 8]
```diff
@@ -39,20 +39,26 @@ public abstract class NoxDatabaseConfigurator : INoxDatabaseConfigurator
 
         public virtual void ConfigureEntity(
             NoxSolutionCodeGeneratorState codeGeneratorState,
-            EntityTypeBuilder builder,
+            IEntityBuilderAdapter builder,
             Entity entity,
-            IReadOnlyList<EntityRelationshipWithType> relationshipsToCreate)
+            NoxSolution noxSolution,
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs:44]: the codeGeneratorState already contains a Solution no need to input it again

File: src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs (hunk 6 above)
Comment (rochar) [reply src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs:13]: the codeGeneratorState contains the function no need to input it here, correct?

File: src/Nox.Types.EntityFramework/Types/HashedText/HashedTextDatabaseConfigurator.cs [hunk 9]
```diff
@@ -13,11 +11,26 @@ public class HashedTextDatabaseConfigurator : INoxTypeDatabaseConfigurator
...
+        Entity entity,
+        bool isKey)
     {
-        builder
-            .OwnsOne(typeof(HashedText), property.Name)
-            .Ignore(nameof(HashedText.Value));
+        var ownedNavigation = builder
+            .OwnsOne(typeof(HashedText), property.Name);
+
+        if (ownedNavigation is EntityTypeBuilder etb)
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Types/HashedText/HashedTextDatabaseConfigurator.cs:24]: why this complexity, this should be hidden behind the adapter?

File: src/Nox.Types.EntityFramework/Types/Money/MoneyDatabaseConfigurator.cs [hunk 10]
```diff
@@ -12,19 +13,30 @@ public class MoneyDatabaseConfigurator : INoxTypeDatabaseConfigurator
...
          var typeOptions = property.MoneyTypeOptions ?? new MoneyTypeOptions();
 
-         builder
-             .OwnsOne(typeof(Money), property.Name)
-             .Ignore(nameof(Money.Value))
-             .Property(nameof(Money.Amount))
-             .IfNotNull(GetColumnType(typeOptions), b => b.HasColumnType(GetColumnType(typeOptions)));
+        var ownedNavigation = builder
+            .OwnsOne(typeof(Money), property.Name);
+        if (ownedNavigation is EntityTypeBuilder etb)
```
Comment (rochar) [inline src/Nox.Types.EntityFramework/Types/Money/MoneyDatabaseConfigurator.cs:26]: the adapter is to avoid this!

**Review by rochar — CHANGES_REQUESTED** [review]

File: src/Nox.Generator/Application/Dto/EntityDto.template.cs (hunk 1 above)
Comment (ILoveMyTortoise) [reply src/Nox.Generator/Application/Dto/EntityDto.template.cs:14]: Errors I receive without it:
Nox.Generator\src\SampleWebApp\Nox.Generator\Nox.Generator.NoxCodeGenerator\Dto.CountryDto.g.cs(110,3,110,13): error CS0246: The type or namespace name 'AutoExpandAttribute' could not be found (are you missing a using directive or an assembly reference?)
Nox.Generator\src\SampleWebApp\Nox.Generator\Nox.Generator.NoxCodeGenerator\Dto.CountryDto.g.cs(110,3,110,13): error CS0246: The type or namespace name 'AutoExpand' could not be found (are you missing a using directive or an assembly reference?)

Comment (ILoveMyTortoise) [reply src/Nox.Generator/Application/Dto/EntityDto.template.cs:14]: Let's discuss, maybe we don't need it

File: src/Nox.Solution/NoxSolution.cs (hunk 2 above)
Comment (ILoveMyTortoise) [reply src/Nox.Solution/NoxSolution.cs:63]: True, but it causes a different issue. Let's discuss the best way on resolving it

File: src/Nox.Solution/NoxSolutionBuilder.cs (hunk 4 above)
Comment (ILoveMyTortoise) [reply src/Nox.Solution/NoxSolutionBuilder.cs:199]: For now yes, but I just wanted to mark this as a place where all similar calculated properties are evaluated

File: src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs (hunk 8 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs:44]: Done

File: src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs (hunk 7 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/Abstractions/NoxDatabaseConfigurator.cs:121]: Done, nice catch

File: src/Nox.Types.EntityFramework/EntityBuilderAdapter/EntityBuilderAdapter.cs
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/EntityBuilderAdapter/EntityBuilderAdapter.cs:1]: Let's discuss this

File: src/Nox.Types.EntityFramework/Types/Money/MoneyDatabaseConfigurator.cs (hunk 10 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/Types/Money/MoneyDatabaseConfigurator.cs:26]: Let's discuss this

File: src/Nox.Types.EntityFramework/Types/HashedText/HashedTextDatabaseConfigurator.cs (hunk 9 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/Types/HashedText/HashedTextDatabaseConfigurator.cs:24]: Let's discuss this

File: src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs (hunk 6 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.EntityFramework/Abstractions/INoxDatabaseConfigurator.cs]: removed it completely

**Review by rochar — APPROVED** [review]

**Pull request merged**